buildDotnetModule: init

IvarWithoutBones opened the pull request:
###### Motivation for this change
This adds an `buildDotnetModule` function, as to avoid code repitition when packaging dotnet applications. This should also make it a lot more beginner friendly, which is always nice. The only thing left to do is adding support for [lockfiles](https://devblogs.microsoft.com/nuget/enable-repeatable-package-restores-using-a-lock-file/) in upstream repositories.  Since Microsoft recommands adding this to your `gitignore`, I couldn't find any projects providing it.

When adding the documentation I noticed the `Dotnet` section never got included in the manual, is there a reason why? I enabled it, hope that is okay.

For now I've only switched over packages I maintain to this new function, but if this get's merged I can switch over the remaining few packages as well. 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@oxalica @kevincox @puffnfresh might be interested in this, seeing how you all maintain dotnet packages.

IvarWithoutBones committed:
buildDotnetModule: init

IvarWithoutBones committed:
ryujinx: use buildDotnetModule

IvarWithoutBones committed:
discordchatexporter-cli: use buildDotnetModule

IvarWithoutBones committed:
buildDotnetModule: add documentation

IvarWithoutBones commented on pkgs/build-support/build-dotnet-module/default.nix:
```diff
@@ -0,0 +1,127 @@
+          makeWrapper "$execPath" "$out/bin/$(basename "$executable")" \
+            --set DOTNET_ROOT "${dotnet-runtime}" \
+            --suffix LD_LIBRARY_PATH : "${lib.makeLibraryPath buildInputs}" \
+            ''${gappsWrapperArgs[@]} \
```
If this is not passed, the argument simply gets ignored. I'm not sure of any other way to do this, without double wrapping when applications do need `wrapGappsHook`.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 139222` at a6f6c814 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>discordchatexporter-cli</li>
    <li>ryujinx</li>
  </ul>
</details>

oxalica commented on pkgs/build-support/build-dotnet-module/default.nix:
```diff
@@ -0,0 +1,127 @@
+
+# Libraries that need to be available at runtime should be passed through this.
+# These get wrapped into `LD_LIBRARY_PATH`.
+, buildInputs ? []
```
We use `buildInputs` for built dependencies everywhere. Using this name for runtime deps is counterintuitive.
I'd suggest `runtimeDeps` or `runtimeDependencies`.

oxalica commented on pkgs/build-support/build-dotnet-module/default.nix:
```diff
@@ -0,0 +1,127 @@
+
+      export HOME=$(mktemp -d)
+      export DOTNET_CLI_TELEMETRY_OPTOUT=1
+      export DOTNET_NOLOGO=1
```
These two `DOTNET_` environment variables can be moved outside as `mkDerivation` arguments.

oxalica commented on pkgs/build-support/build-dotnet-module/default.nix:
```diff
@@ -0,0 +1,127 @@
+      mkdir -p $HOME/.nuget/NuGet
+      cp $HOME/.config/NuGet/NuGet.Config $HOME/.nuget/NuGet
+
+      dotnet restore --source "$PWD/nixos" ${projectFile} ''${dotnetRestoreFlags} ''${dotnetFlags}
```
I assume `projectFile` is the relative path to the project file, which is from upstream. I think we should escape them with `lib.escapeShellArg`.

kevincox replied:
The arguments should also be quoted in the script.

```suggestion
      dotnet restore --source "$PWD/nixos" ${projectFile} "''${dotnetRestoreFlag[@]}" "''${dotnetFlags[@]}"
```

And the same throughout.

oxalica commented on pkgs/misc/emulators/ryujinx/default.nix:
```diff
@@ -28,70 +15,40 @@ in stdenv.mkDerivation rec {
-    dotnet publish Ryujinx.sln \
-      --no-build \
-      --configuration Release \
-      --no-self-contained \
```
`--no-self-contained` is removed and so does `discordchatexporter-cli`. Does it matter? Would it cause dotnet runtime to be copied to $out which bloats output size?

If this option is required, maybe we need to change the default wrapper and set them in default `dotnetFlags`.

IvarWithoutBones replied:
> `--no-self-contained` is removed and so does `discordchatexporter-cli`. Does it matter? Would it cause dotnet runtime to be copied to $out which bloats output size?

I think you're right about the closure bloat, somehow completely missed the flag. I'll add it back to the default wrapper as you suggested :+1:

oxalica commented on pkgs/tools/backup/discordchatexporter-cli/default.nix:
```diff
@@ -16,65 +17,12 @@ stdenv.mkDerivation rec {
-    runHook postInstall
-  '';
+  projectFile = "DiscordChatExporter.Cli/DiscordChatExporter.Cli.csproj";
+  dotnet-runtime = dotnetCorePackages.aspnetcore_3_1;
```
Why it's changed from `sdk_5` to `aspnetcore_3_1`?
And can `dotnet-sdk` and `dotnet-runtime` have different version? The former is not overrided here.

IvarWithoutBones replied:
> Why it's changed from sdk_5 to aspnetcore_3_1?

I guess I didn't pay enough attention when converting this package, it should be `dotnetCorePackages.netcore_3_1`. Changing it to that :)

> And can dotnet-sdk and dotnet-runtime have different version? The former is not overrided here.

Yes, this is actually the only way to use this package, as per the [project's readme](https://github.com/Tyrrrz/DiscordChatExporter#building-the-project-locally).

oxalica commented on doc/languages-frameworks/dotnet.section.md:
```diff
@@ -70,6 +70,39 @@ The `dotnetCorePackages.sdk` contains both a runtime and the full sdk of a given
-Ideally, we would like to build against the sdk, then only have the dotnet runtime available in the runtime closure.
+To package Dotnet applications, you can use `buildDotnetModule`. This has similar arguments to `stdenv.mkDerivation`, with the following additions:
+
+* `projectFile` has to be used for specifying the dotnet project file. These usually have `.sln` or `.csproj` file extensions.
```
```suggestion
* `projectFile` has to be used for specifying the dotnet project file relative to the source root. These usually have `.sln` or `.csproj` file extensions.
```

oxalica commented:
For the issue of duplicated large generated files, using upstream lock file is indeed the easiest way. But to remind, we can also use collect them and generate dependencies in a single run to a single file like what we did with `nodePackages`, which shares common dependencies.
But of course, this PR is the first step to abstract this building process.

kevincox commented on pkgs/build-support/build-dotnet-module/default.nix:
```diff
@@ -0,0 +1,127 @@
+, dotnet-sdk ? dotnetCorePackages.sdk_5_0
+, ... } @ args:
+
+assert projectFile == null -> throw "Defining the `projectFile` attribute is required. This is usually an `.csproj`, or `.sln` file.";
+
+# TODO: Automatically generate a dependency file when a lockfile is present.
+# This file is unfortunately almost never present, as Microsoft recommands not to push this in upstream repositories.
+assert nugetDeps == null -> throw "Defining the `nugetDeps` attribute is required, as to lock the NuGet dependencies. This file can be generated using the `nuget-to-nix` tool.";
```
Why not just leave these without a default?

IvarWithoutBones replied:
I thought this would be nicer, as this way we can create clearer error messages. If not setting a default is prefered, I'll change it to that.

kevincox reviewed (commented):
Overall it looks good but I know very little about dotnet.

cc @jojosch who actually moves Jellyfin to building from source.

IvarWithoutBones commented:
Thanks for the reviews everyone! I just pushed the suggestions.

> For the issue of duplicated large generated files, using upstream lock file is indeed the easiest way. But to remind, we can also use collect them and generate dependencies in a single run to a single file like what we did with nodePackages, which shares common dependencies.

That's a good point, would definitly be very useful to have. If possible, i would like to generate this automatically if `nugetDeps` is undefined. I'll look into this in the future, though for now I think this should be fine.

IvarWithoutBones commented:
I have also just added support for disabling multi threaded compilation, and added the [`Deterministic` flag](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/code-generation#deterministic).

oxalica reviewed (changes requested):
I build `ryujinx` with `nix-build . -A ryujinx --option repeat 3` and it failed to produce identical output. Well, it's also non-deterministic on master so this issue may not be related to this PR.

IvarWithoutBones commented:
Added the suggested changes :+1: 

> I build ryujinx with nix-build . -A ryujinx --option repeat 3 and it failed to produce identical output. Well, it's also non-deterministic on master so this issue may not be related to this PR.

Huh, that's odd, I thought it was deterministic. I'll look into this, though for now I think this PR be fine seeing how it happens on master as well.

IvarWithoutBones commented:
/marvin opt-in
/status needs_reviewer

SuperSandro2000 commented on pkgs/build-support/build-dotnet-module/default.nix:
```diff
@@ -0,0 +1,144 @@
+      cp $HOME/.config/NuGet/NuGet.Config $HOME/.nuget/NuGet
+
+      dotnet restore ${lib.escapeShellArg projectFile} \
+        ${if !enableParallelBuilding then "--disable-parallel" else ""} \
```
```suggestion
        ${lib.optionalString (!enableParallelBuilding) "--disable-parallel"} \
```

SuperSandro2000 commented on pkgs/build-support/build-dotnet-module/default.nix:
```diff
@@ -0,0 +1,144 @@
+      done
+    '' else ''
+      for executable in $out/lib/${args.pname}/*; do
+        if [[ -f "$executable" ]] && [[ -x "$executable" ]] && [[ "$executable" != *"dll"* ]]; then
```
```suggestion
        if [[ -f "$executable" && -x "$executable" && "$executable" != *"dll"* ]]; then
```

IvarWithoutBones commented:
Adressed all feedback :+1:
